### ISSUE TITLE
(Tracking Issue) Ubernetes (Federation) Lite

### ISSUE THREAD

**Issue by justinsb** [kubernetes/kubernetes#17059]
This issue is to track progress on Ubernetes (Federation) Lite.

Ubernetes Lite is a single k8s cluster running across multiple availability zones (AWS) or zones (GCE).  Other clouds & bare metal will also be possible, but won't be implemented as part of this proposal.  (But much of the infrastructure should be the same, e.g. for bare metal, if you tag nodes & volumes correctly the scheduler constraints should "just work")

I have this working on my branch, and I am cleaning up code and PRing chunks.  The 'end result' can be seen on my branch:  https://github.com/justinsb/kubernetes/tree/multizone_with_labels

There are various elements to the plan (this is the AWS track):

- [x] MVP changes to kube-up so that it can create nodes in multiple zones for AWS (MVP done in #16735, better support depends on e.g. #16199)
- [x] Make any needed changes to the AWS cloud provider (believed to be none)
- [x] Change kubelet to label new nodes with Zone information (#16866)
- [x] Change the scheduler to be spread across zones as a preference (PoC https://github.com/justinsb/kubernetes/commit/d8d8ec2b548abf040a93db52a8c9012b5d7fe20d).
- [x] Add zone labels to persistent volumes (using an admission controller?  PoC https://github.com/justinsb/kubernetes/commit/8c2a87a3a183ad9c2aeee7e6ae7cfc0a979fe2f6)  Likely only for PersistentVolumes initially.
[ ] Use volume zone labels as a hard-constraint on pod scheduling (PoC https://github.com/justinsb/kubernetes/commit/670c033b2cf85e49fedaa226f0597f80c752719f)

In parallel we do GCE:

- [x] Change kube-up so that it can create nodes in multiple zones for GCE (PoC of MVP level: https://github.com/justinsb/kubernetes/commit/173b17aeba21c7cb59a65593e01839d0b852993b)
- [x] Change GCE cloud provider to be multi-zone aware.  It looks like this will likely require a feature flag because of changes to the semantics of zone awareness (PoC https://github.com/justinsb/kubernetes/commit/f54943bdf873041af8dd969804ab69bc4d67aa72)

That is the intended initial scope of the initial Ubernetes Lite work.  In particular, we're not planning on implementing things like HA master or zone awareness in the kube-proxy as part of this effort.  But, if HA master is implemented we can enhance it to make it multi-zone.  Or if kube-proxy adds node-locality routing, then maybe we can add zone-locality routing also.  Otherwise, the answer for more advanced functionality is likely to be "that will be done in Ubernetes Proper".

**Comment by davidopp** [comment]
@kevin-wangzefeng says:

Very attractive feature. Our customers have some similar requirements:  

1. Multi-AZ nodes in one k8s cluster.  (also found in some previous discussion #4235 #13056 #13063)
    - User can choose which one or more AZ(s) to run Pods in.
    - Pods belongs to one RC will be spread across zones as a preference  
    - Pods (belongs to different services) that talks to each other, will be running in the same AZ.

2. Run one  k8s cluster across multiple VLANs
    - Users will specify in which VLAN they want Pods to run.
    - Also some customized scheduling policies based on configuration.

In a summary, the common points (including this issue) are:
- partition nodes into groups 
- add some new scheduling policies based on these groups.

We've just started to do the design work, and I have similar views on the implementation as you mentioned above, which is a easy way to start.

A quick thought about labeling each node to denote a group name, we can make it configurable, or define some well known label key like this way [https://github.com/justinsb/kubernetes/commit/5737b1fb551ec0a72246b49328e3d4f88a3d5a70].

------------------------------------------------------

@davidopp replies:

For (1)
* first bullet: this is not possible yet but will be possible when we implement NodeSelector which will have an "or" expression ( https://github.com/kubernetes/kubernetes/issues/341#issuecomment-154663349 )
* second bullet: you can use the scheduling priority function ServiceAntiAffinityPriority with label key of "az" or whatever key you use to label each node with which AZ it is in
* third bullet: we don't currently support that; I have some questions about that but let's discuss it in the other issue

For (2)
* first bullet: I believe this can be handled today using node labels and label selectors (the label could be added manually or we could add code to automatically add it like https://github.com/justinsb/kubernetes/commit/5737b1fb551ec0a72246b49328e3d4f88a3d5a70 though I think manual is simpler for now)
* second bullet: this should be possible by writing your own scheduling algorithm or, in the worst case, your own scheduler (#17197)

**Comment by alfred-huangjian** [comment]
@davidopp @justinsb @kevin-wangzefeng 
For(1), 
It make sense for the first bullet and second bullet.
third bullet: we prefer to reduce the network communications between zones. 
1st. Related applications can be deployed in the same several zones if possible. Just like Web application access the database.
2nd. Give the high priorities in kube-proxy routing iptables to visit the service in the same zone or even the same node.

For (2), we can not use the label selector because first the pod can be arranged anywhere at first.  But after the node is chooses, it can't move out the scoped of the nodes. Currently, we are writing our own scheduler to solve the specific issue. But prefer to find some general solutions if possible.

**Comment by davidopp** [comment]
> For(1), 
It make sense for the first bullet and second bullet.
third bullet: we prefer to reduce the network communications between zones. 
1st. Related applications can be deployed in the same several zones if possible. Just like Web application access the database.
2nd. Give the high priorities in kube-proxy routing iptables to visit the service in the same zone or even the same node.

I believe what you're asking for is #15675. Is that correct?

> For (2), we can not use the label selector because first the pod can be arranged anywhere at first. But after the node is chooses, it can't move out the scoped of the nodes. Currently, we are writing our own scheduler to solve the specific issue. But prefer to find some general solutions if possible.

Can you add this request to #14484? I think that is the most relevant issue (and I believe it applies for any label key, not just a label that specifies the zone in a multi-AZ configuration).

**Comment by brendandburns** [comment]
@justinsb I believe all items in this tracking bug are completed.  Shall we close this?

**Comment by justinsb** [comment]
I think so @brendandburns :-)

I believe that #15675 tracks the idea that we should prefer to connect to a pod on the same node or same zone (we now have label information for zones, so it's much more practical but this was deliberately out of scope for Ubernetes-Lite at least in 1.2); it sounds like #14484 is the right place to discuss RC affinity.

On Ubernetes-Lite itself, the last PR is #19995 and I think it's pretty close to merging.  I have started (emphasis on started) some docs in #20044 and I'm sure we'll find a few things that need fixing when we put this through our full test suites, but I think this issue is resolved.

**Comment by davidopp** [comment]
>  it sounds like #14484 is the right place to discuss RC affinity.

actually #18265

**Comment by davidopp** [comment]
Can someone link to the PRs that implemented each of the things in the first entry in this thread? That will be super-useful for someone who comes along later and wants to see what are the pieces that went into Ubernetes Lite.

**Comment by pires** [comment]
I second @davidopp comment.

**Comment by justinsb** [comment]
Sure...  here are the primary ones:

AWS: Experiment support for multiple AZs / subnets #16735
Label nodes with Zone information, if available #16866
Make zone and region node labels consistent with our convention #17506
Ubernetes: When scheduling, spread between zones if labeled #17915
Ubernetes Lite: Volumes can dictate zone scheduling #17916
Ubernetes Lite support for GCE #17919
Design doc for Ubernetes Lite #19257
Ubernetes-Lite GCE: Label volumes with zone information #19995

And help-wanted on:
WIP docs for ubernetes-lite #20044

Once I get AWS passing e2e and federating build results in (soon!) I'll attack those docs again...

**Comment by davidopp** [comment]
I haven't looked at the Ubernetes Lite code carefully, but does it do replicated master components, or is the master just running in one zone? If the latter, I think an important follow-on would be to do replicated master components so a zone failure isn't a SPOF. (After all, one of the main reasons people will run Ubernetes Lite is to avoid a zone failure being a SPOF.)

**Comment by davidopp** [comment]
Re-opening until question I asked above is answered.

**Comment by justinsb** [comment]
It doesn't do HA master; we still have a single master.  There's a fair amount of work to be done to have true HA master.  There's still value, in that zone failure won't take out your app, though it will take out your control plane in 1.2.

We do have auto-restart on AWS (and it was already on GCE), so that if the master instance crashes it should be restarted by CloudWatch alarms.

I'm hoping we'll get true HA in 1.3 - there's a SIG dedicated to it, I believe!

**Comment by davidopp** [comment]
Yes, you make a very valid point (that since the app will continue running even if the master goes down, you do make your app significantly more available using the current Ubernetes Lite implementation even though it doesn't do HA master). Anyway, I'll file a separate issue for that and close this one. Thanks!
